Check for missing app code before upgrade

VicDeo opened the pull request:
## Description


## Related Issue
 #26227

## Motivation and Context
See issue

## How Has This Been Tested?


## Screenshots (if appropriate):

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

VicDeo committed:
Check for missing app code before upgrade #26227

VicDeo committed:
Add an explanation. Propagate check to the web updater

VicDeo commented:
@PVince81 what do you think?

DeepDiver1975 commented on core/ajax/update.php:
```diff
@@ -91,6 +91,18 @@ public function handleRepairFeedback($event) {
+	$preUpdate = new \OC\Updater\PreUpdate(\OC::$server->getAppManager());
+	$missingApps = $preUpdate->getMissingApps();
+	if (count($missingApps) !== 0){
+		$eventSource->send('notice', (string)$l->t('Code is missing for the following apps:'));
```
How does this look like in the web ui? Please add screenshot

jvillafanez replied:
Any reason why this is being converted to a string? It should be returning a string already 😕

VicDeo replied:
@jvillafanez dunno, just like all the rest of `$l->t()` invocations in this file

DeepDiver1975 commented on core/ajax/update.php:
```diff
@@ -91,6 +91,18 @@ public function handleRepairFeedback($event) {
+		foreach ($missingApps as $appId){
+			$eventSource->send('notice', $appId);
+		}
+		$eventSource->send('failure', (string)$l->t('Please use occ app:disable command if you don\'t need these apps or restore the code.'));
```
Kind of hard from the web ui ....

DeepDiver1975 replied:
my comment remains valid I guess .... what about auto-disabling them when accessing via web? or in a second request?

VicDeo replied:
@DeepDiver1975 ok, will do

DeepDiver1975 commented on tests/Core/Command/Upgrade/CheckAppsTest.php:
```diff
@@ -0,0 +1,89 @@
+/**
+ * @author Victor Dubiniuk <dubiniuk@owncloud.com>
+ *
+ * @copyright Copyright (c) 2017, ownCloud GmbH.
```
remove the dot please

VicDeo commented:
@DeepDiver1975 
![screenshot_20170130_223744](https://cloud.githubusercontent.com/assets/991300/22438407/ed4bcd6a-e73c-11e6-9ff6-95dae1575258.png)

jvillafanez commented on core/Command/Upgrade/CheckApps.php:
```diff
@@ -0,0 +1,80 @@
+		$this
+			->setName('upgrade:checkapps')
+			->setDescription('Check if there are enabled apps with missing code.')
+		;
```
I'd move this up. It looks weird.

VicDeo replied:
@jvillafanez sorry, what's exactly wrong here?

jvillafanez replied:
It's code style. Having a `;` isolated there is weird.

VicDeo replied:
@jvillafanez thanks for clarification, updated.

jvillafanez commented on core/Command/Upgrade/CheckApps.php:
```diff
@@ -0,0 +1,80 @@
+			if ($input->getOption('output') === self::OUTPUT_FORMAT_PLAIN){
+				$output->writeln('Code is missing for the following apps:');
+				$this->writeArrayInOutputFormat($input, $output, $missingApps);
+				$output->writeln('Please use occ app:disable command if you don\'t need these apps or restore the code.');
```
I trust this works fine. I wouldn't like that the command the users is supposed to execute to fix the situation fails. Please confirm.

VicDeo replied:
```
deo@jah-mobile:~/public_html/oc-tmp> php occ app:list
Enabled:
  - admin_audit: 0.7
  - comments: 0.3.0
  - dav: 0.2.8
  - enterprise_key: true
  - federatedfilesharing: 0.3.0
  - federation: 0.1.0
  - files: 1.5.1
  - files_sharing: 0.10.0
  - files_versions: 1.3.0
  - gallery: 15.0.0
  - provisioning_api: 0.5.0
  - systemtags: 0.3.0
  - updatenotification: 0.2.1
Disabled:
  - activity
  - encryption
  - files_antivirus
  - files_drop
  - files_external
  - files_texteditor
  - files_trashbin
  - files_videoviewer
  - richdocuments
  - templateeditor
  - testing

deo@jah-mobile:~/public_html/oc-tmp> mv apps/admin_audit/ ../admin_audit
deo@jah-mobile:~/public_html/oc-tmp> php occ upgrade:checkapps
Code is missing for the following apps:
  - admin_audit
Please use occ app:disable command if you don't need these apps or restore the code.

deo@jah-mobile:~/public_html/oc-tmp> php occ app:disable admin_audit
admin_audit disabled
deo@jah-mobile:~/public_html/oc-tmp> php occ upgrade:checkapps
deo@jah-mobile:~/public_html/oc-tmp>
```

jvillafanez replied:
👍

jvillafanez commented on core/Command/Upgrade/CheckApps.php:
```diff
@@ -0,0 +1,80 @@
+				$this->writeArrayInOutputFormat($input, $output, [ 'missing' => $missingApps ]);
+			}
+
+
```
Unneeded empty lines.

jvillafanez commented on lib/private/Updater/PreUpdate.php:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	/**
+	 * @return array
```
Taking into account this is a public method, the return value should be clearer. Is it returning a list of strings? a list of numbers? a list of app objects? an array with a specific format such as `[<appid> => true]`?

jvillafanez commented on lib/private/Updater/PreUpdate.php:
```diff
@@ -0,0 +1,48 @@
+
+use OCP\App\IAppManager;
+
+class PreUpdate {
```
I don't have a clear idea what is the porpuse of this class. Why is this class needed? Why we can't move the `getMissingApps` method inside the AppManager, or in any other class to manage apps?

VicDeo replied:
@jvillafanez This check is done before upgrade. And there will be more checks in future.

jvillafanez replied:
Maybe an interface `IPreUpgradeCheck` with a `doCheck` method (not sure about what parameters should be included) might be a better choice.
For a specific check you can implement the interface and do whatever you want to do inside the `doCheck` method.

The only potential problem would be the implementation discovery. For now, core could create the implementations he's willing to go through during the upgrade process.

@DeepDiver1975 does it make sense?

DeepDiver1975 replied:
> This check is done before upgrade. And there will be more checks in future.

please add this to the classÄ phpdoc

DeepDiver1975 replied:
> Maybe an interface IPreUpgradeCheck with a doCheck method (not sure about what parameters should be included) might be a better choice.

since we do not expect any public api usage it is okay to it this way

DeepDiver1975 commented on lib/private/Updater/PreUpdate.php:
```diff
@@ -0,0 +1,48 @@
+		$missingApps = [];
+		foreach ($installedApps as $appId){
+			$info = $this->appManager->getAppInfo($appId);
+			if (!isset($info['id'])){
```
I guess you should better test for null

VicDeo replied:
@DeepDiver1975 May be both.  What if there is no key `id`?

VicDeo committed:
Try to autodisable missing app when running from web

VicDeo commented:
@DeepDiver1975 updated, rebased, squashed.
CLI still offer to take any action on their own

Web UI autodisables the app:
![new_screen](https://cloud.githubusercontent.com/assets/991300/24308049/8c342f62-10d7-11e7-9ef9-d589fe9cc675.png)

VicDeo commented:
@DeepDiver1975 so, is it good enough now?

PVince81 commented:
Tested, works.

However in the web UI I think the update should abort if apps have missing code.

The reason is that if the app that is missing has some important routines like user backends, disabling that app is likely to break things because some migration code iterates over ALL users. But in the case of a missing user backend, some users would not be migrated at all.

The CLI behavior is fine.

PVince81 commented:
and then comes the question: what about people who don't have CLI access, how would they disable these apps manually ? Hmmm... maybe the updater needs to show an error and then provide a button "click here to disble the apps" which reloads the page with an override URL parameter.

thoughts ?

VicDeo commented:
@PVince81 as suggested by @DeepDiver1975 it is autodisabled in web UI. (See the last commit)

PVince81 commented:
The reason I asked for this feature was specifically to avoid migrations that forgot users due to disabled app. If we auto disable apps then we lose this and people will still result in broken envs.

I could imagine this happening in case there are bugs in the "bundled app to market migration" code where for example user_ldap used to be on-disk (repo package) but now in the new version it's not there but must be downloaded from the market. But then if download failed and the upgrade continues (we likely don't want to continue here!), then user_ldap app is missing and the migration code of other apps that need all users will not find the LDAP users.

@DeepDiver1975

PVince81 commented:
of course user_ldap disallowes running web UI update, but some other user backends might

PVince81 commented:
This becomes even more important when unbundling apps to the marketplace. The risk of an app's code missing is much higher.

PVince81 commented:
This becomes even more important when unbundling apps to the marketplace. The risk of an app's code missing is much higher.

PVince81 commented:
- [ ] reuse this PR for https://github.com/owncloud/market/issues/39
- [ ] make sure the web UI also aborts (for now we provide no option to disable apps from the UI, something for 10.0.1 possibly using the "secret" "disable app short-circuit API even when OC is broken" endpoint)

PVince81 commented:
Obsoleted by https://github.com/owncloud/core/pull/27711
